src/librustc_typeck/check/generator_interior.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use rustc::hir::intravisit::{self, Visitor, NestedVisitorMap};
+use rustc::hir::{self, Body, Pat, PatKind, Expr};
+use rustc::hir::def_id::DefId;
+use rustc::ty::Ty;
+use rustc::middle::region::{RegionMaps, CodeExtent, extent_has_yield};
+use util::nodemap::FxHashSet;
+use std::rc::Rc;
+use super::FnCtxt;
+
+struct InteriorVisitor<'a, 'gcx: 'a+'tcx, 'tcx: 'a> {
+    fcx: &'a FnCtxt<'a, 'gcx, 'tcx>,
+    types: FxHashSet<Ty<'tcx>>,
+    region_maps: Rc<RegionMaps>,
+}
+
+impl<'a, 'gcx, 'tcx> InteriorVisitor<'a, 'gcx, 'tcx> {
+    fn record(&mut self, ty: Ty<'tcx>, scope: Option<CodeExtent>, expr: Option<&'tcx Expr>) {
+        use syntax_pos::DUMMY_SP;
+
+        if scope.map(|s| extent_has_yield(self.fcx.tcx, s)).unwrap_or(true) {
+            if self.fcx.tcx.sess.verbose() {
+                if let Some(s) = scope {
+                    self.fcx.tcx.sess.span_warn(s.span(&self.fcx.tcx.hir).unwrap_or(DUMMY_SP),
+                        &format!("type in generator with scope = {:?}, type = {:?}",
+                                 scope,
+                                 self.fcx.resolve_type_vars_if_possible(&ty)));
+                } else {
+                    self.fcx.tcx.sess.span_warn(DUMMY_SP,
+                        &format!("type in generator WITHOUT scope, type = {:?}",
+                                 self.fcx.resolve_type_vars_if_possible(&ty)));
+                }
+                if let Some(e) = expr {
+                    self.fcx.tcx.sess.span_warn(e.span,
+                        &format!("type from expression: {:?}", e));
+                }
+            }
+            self.types.insert(ty);
+        } else if self.fcx.tcx.sess.verbose() {
+            if let Some(e) = expr {
+                self.fcx.tcx.sess.span_warn(e.span,
+                    &format!("NO type from expression: {:?}", e));
+            }
+        }
+    }
+}
+
+pub fn find_interior<'a, 'gcx, 'tcx>(fcx: &'a FnCtxt<'a, 'gcx, 'tcx>,
+                                     def_id: DefId,
+                                     body_id: hir::BodyId,
+                                     witness: Ty<'tcx>) {
+    let body = fcx.tcx.hir.body(body_id);
+    let mut visitor = InteriorVisitor {
+        fcx,
+        types: FxHashSet(),
+        region_maps: fcx.tcx.region_maps(def_id),
+    };
+    intravisit::walk_body(&mut visitor, body);
+
+    // FIXME: Drop elaboration can insert bool types in the generator
```

Comment:
I don't think we should expose the fact that drop elaboration happens to use `bool` to users.